app/modules/softuart.c

```diff
@@ -0,0 +1,398 @@
+    } else {
+        //clear interrupt, no matter from which pin
+        //otherwise, this interrupt will be called again forever
+        GPIO_REG_WRITE(GPIO_STATUS_W1TC_ADDRESS, gpio_status);
+    }
+
+    ret_gpio_status &= ~(mask);
+
+    return ret_gpio_status;
+}
+
+void softuart_init()
+{
+    Softuart* s = &softuart;
+
+    NODE_DBG("SOFTUART initialize gpio\r\n");
+    gpio_init();
+
+	//set bit time
+    uint32_t baudrate = s->baudrate;
+    s->bit_time = (1000000 / baudrate);
+    if (((100000000 / baudrate) - (100*s->bit_time)) > 50 ) s->bit_time++;
+    NODE_DBG("SOFTUART bit_time is %d\r\n", s->bit_time);
+
+	//init tx pin
+    platform_gpio_mode(s->pin_rx, PLATFORM_GPIO_OUTPUT, PLATFORM_GPIO_PULLUP);
+    GPIO_OUTPUT_SET(GPIO_ID_PIN(s->pin_tx), 1);
+    uint32_t delay = 100000; //TODO needed?
+    os_delay_us(delay);
+    NODE_DBG("SOFTUART TX INIT DONE\r\n");
+
+    //init rx pin
+    platform_gpio_mode(s->pin_rx, PLATFORM_GPIO_INT, PLATFORM_GPIO_PULLUP);
+    uint32_t mask = 1 << s->pin_rx;
+    platform_gpio_register_intr_hook(mask, softuart_intr_handler);
+
+    //clear interrupt handler status, basically writing a low to the output
+    GPIO_REG_WRITE(GPIO_STATUS_W1TC_ADDRESS, BIT(s->pin_rx)); //TODO needed?
+
+    //enable interrupt for pin on any edge (rise and fall)
+    gpio_pin_intr_state_set(GPIO_ID_PIN(s->pin_rx), GPIO_PIN_INTR_ANYEDGE);
+    NODE_DBG("SOFTUART RX INIT DONE\r\n");
+}
+
+static inline u8 chbit(u8 data, u8 bit)
+{
+    if ((data & bit) != 0)
+    {
+    	return 1;
+    }
+    else
+    {
+    	return 0;
+    }
+}
+
+// Function for printing individual characters
+void softuart_putchar(char data)
+{
+    NODE_DBG("putchar called\r\n");
+    Softuart* s = &softuart;
+	unsigned i;
+	unsigned start_time = 0x7FFFFFFF & system_get_time();
+
+	//Start Bit
+	GPIO_OUTPUT_SET(GPIO_ID_PIN(s->pin_tx), 0); //TODO replay by plaform function
+	for(i = 0; i < 8; i ++ )
+	{
+		while ((0x7FFFFFFF & system_get_time()) < (start_time + (s->bit_time*(i+1))))
+		{
+			//If system timer overflow, escape from while loop
+			if ((0x7FFFFFFF & system_get_time()) < start_time){break;}
+		}
+		GPIO_OUTPUT_SET(GPIO_ID_PIN(s->pin_tx), chbit(data,1<<i));
+	}
+
+	// Stop bit
+	while ((0x7FFFFFFF & system_get_time()) < (start_time + (s->bit_time*9)))
+	{
+		//If system timer overflow, escape from while loop
+		if ((0x7FFFFFFF & system_get_time()) < start_time){break;}
+	}
+	GPIO_OUTPUT_SET(GPIO_ID_PIN(s->pin_tx), 1);
+
+	// Delay after byte, for new sync
+	os_delay_us(s->bit_time*6);
+}
+
+
+static int softuart_setup(lua_State* L)
+{
+    uint32_t result = 1;
+    uint32_t baudrate = (uint32_t) luaL_checkinteger(L, 1);
+    uint8_t tx_gpio_id = (uint8_t) luaL_checkinteger(L, 2);
+    uint8_t rx_gpio_id = (uint8_t) luaL_checkinteger(L, 3);
+
+    if(softuart.enabled)
+        return luaL_error( L, "SOFTUART already configured." );
+
+    if(!softuart_gpio_valid(tx_gpio_id)) {
+        return luaL_error(L, "SOFTUART tx GPIO not valid");
```

Comment:
Hi,

It would be beneficial to allow RX or TX only usage by accepting NIL as RX/TX pin. Of course in that case you cannot use read/write methods.